components/layout/display_list/builder.rs

```diff
@@ -3198,3 +3208,42 @@ pub struct BackgroundPlacement {
     /// measures above it is only shown within these bounds.
     css_clip: Rect<Au>,
 }
+
+pub struct IndexableTextItem {
+    /// The placement of the text item on the plane.
+    pub origin: Point2D<Au>,
+    /// The text run.
+    pub text_run: Arc<TextRun>,
+    /// The range of text within the text run.
+    pub range: Range<ByteIndex>,
+    /// The position of the start of the baseline of this text.
+    pub baseline_origin: Point2D<Au>,
+}
+
+#[derive(Default)]
+pub struct IndexableText {
+    inner: FnvHashMap<OpaqueNode, IndexableTextItem>,
+}
+
+impl IndexableText {
+    fn insert(&mut self, node: OpaqueNode, item: IndexableTextItem) {
+        // Guard against duplicate nodes.
+        if self.inner.insert(node, item).is_some() {
+            debug!(
+                "TODO(#20020): Text indexing for {:?} is broken because of multiple text runs.",
```

Comment:
Hmm... using a `HashMap<OpaqueNode, IndexableTextItem>` might make it harder to fix #20020.  In the current code, all the data is in the display list, but the loop incorrectly stops at the first display item for the node.  It seems like we could fix #20020 by continuing to search until we find the text run that actually contains the given advance.  With the hash map, we don't even store all the text runs.

This doesn't need to be fixed right now, but what is the long-term plan?  Maybe the values in the HashMap will change to something like `SmallVec<[IndexableTextItem; 1]>`?

Reply:
I just realized that this also impacts #19754 which adds new code that queries for the text runs.

Reply:
This doesn't need a full fix for #20020, but please at least change it to store all the text, rather than just the first fragment/run of each node, so that #19754 is still possible.  Thanks!